File: reactive-sql-client-examples/src/main/java/io/vertx/example/sqlclient/transaction/SqlClientExample.java
package io.vertx.example.sqlclient.transaction;

import io.vertx.core.AbstractVerticle;
import io.vertx.example.util.Runner;
import io.vertx.pgclient.PgConnectOptions;
import io.vertx.pgclient.PgPool;
import io.vertx.sqlclient.Pool;
import io.vertx.sqlclient.PoolOptions;
import io.vertx.sqlclient.Row;

import java.util.ArrayList;
import java.util.List;

/*
 * @author <a href="mailto:pmlopes@gmail.com">Paulo Lopes</a>
 */
public class SqlClientExample extends AbstractVerticle {

  // Convenience method so you can run it in your IDE
  public static void main(String[] args) {
    Runner.runExample(SqlClientExample.class);
  }

  @Override
  public void start() {

    Pool pool = PgPool.pool(vertx, new PgConnectOptions()
      .setPort(5432)
      .setHost("the-host")
      .setDatabase("the-db")
      .setUser("user")
      .setPassword("secret"), new PoolOptions().setMaxSize(4));

    // Uncomment for MySQL
//    Pool pool = MySQLPool.pool(vertx, new MySQLConnectOptions()
//      .setPort(5432)
//      .setHost("the-host")
//      .setDatabase("the-db")
//      .setUser("user")
//      .setPassword("secret"), new PoolOptions().setMaxSize(4));

    pool.withTransaction(sqlClient -> {
      // create a test table
      return sqlClient.query("create table test(id int primary key, name varchar(255))").execute()
        .compose(v -> {
          // insert some test data
          return sqlClient.query("insert into test values (1, 'Hello'), (2, 'World')").execute();
        })
        .compose(v -> {
          // query some data
          return sqlClient.query("select * from test").execute();
        })
        .map(rows -> {
          // convert rows to strings
          List<String> res = new ArrayList<>(rows.size());
          for (Row row : rows) {
            res.add(row.toString());
          }
          return res;
        });
    }).onComplete(ar -> {
      if (ar.succeeded()) {
        List<String> rows = ar.result();
        System.out.println("rows = " + rows);
      } else {
        ar.cause().printStackTrace();
      }
    });
  }
}

File: reactive-sql-client-examples/src/main/kotlin/io/vertx/example/sqlclient/transaction/SqlClientExample.kt
package io.vertx.example.sqlclient.transaction

import io.vertx.kotlin.pgclient.*
import io.vertx.kotlin.sqlclient.*
import io.vertx.pgclient.PgConnectOptions
import io.vertx.pgclient.PgPool
import io.vertx.sqlclient.PoolOptions

class SqlClientExample : io.vertx.core.AbstractVerticle()  {
  override fun start() {

    var pool = PgPool.pool(vertx, PgConnectOptions(
      port = 5432,
      host = "the-host",
      database = "the-db",
      user = "user",
      password = "secret"), PoolOptions(
      maxSize = 4))

    // Uncomment for MySQL
    //    Pool pool = MySQLPool.pool(vertx, new MySQLConnectOptions()
    //      .setPort(5432)
    //      .setHost("the-host")
    //      .setDatabase("the-db")
    //      .setUser("user")
    //      .setPassword("secret"), new PoolOptions().setMaxSize(4));

    pool.begin({ res1 ->
      if (res1.failed()) {
        System.err.println(res1.cause().getMessage())
        return
      }
      var tx = res1.result()

      // create a test table
      tx.query("create table test(id int primary key, name varchar(255))", { res2 ->
        if (res2.failed()) {
          tx.close()
          System.err.println("Cannot create the table")
          res2.cause().printStackTrace()
          return
        }

        // insert some test data
        tx.query("insert into test values (1, 'Hello'), (2, 'World')", { res3 ->

          // query some data with arguments
          tx.query("select * from test", { rs ->
            if (rs.failed()) {
              System.err.println("Cannot retrieve the data from the database")
              rs.cause().printStackTrace()
              return
            }

            for (line in rs.result()) {
              println("${line}")
            }

            // and close the connection
            tx.commit()
          })
        })
      })
    })
  }
}
